engine-precompiles/src/xcc.rs

```diff
@@ -160,16 +223,102 @@ impl<I: IO> Precompile for CrossContractCall<I> {
+    use borsh::{BorshDeserialize, BorshSerialize};
+
+    pub const ERR_CORRUPTED_STORAGE: &str = "ERR_CORRUPTED_XCC_STORAGE";
+    pub const ERR_MISSING_WNEAR_ADDRESS: &str = "ERR_MISSING_WNEAR_ADDRESS";
+    pub const VERSION_KEY: &[u8] = b"version";
+    pub const WNEAR_KEY: &[u8] = b"wnear";
+    /// Amount of NEAR needed to cover storage for a router contract.
+    pub const STORAGE_AMOUNT: Yocto = Yocto::new(2_000_000_000_000_000_000_000_000);
+
+    /// Type wrapper for version of router contracts.
+    #[derive(
+        Debug,
+        Clone,
+        Copy,
+        Default,
+        PartialEq,
+        Eq,
+        PartialOrd,
+        Ord,
+        BorshDeserialize,
+        BorshSerialize,
+    )]
+    pub struct CodeVersion(pub u32);
+
+    impl CodeVersion {
+        pub fn increment(self) -> Self {
+            Self(self.0 + 1)
+        }
+    }
+
+    /// Get the address of the wNEAR ERC-20 contract
+    pub fn get_wnear_address<I: IO>(io: &I) -> Address {
+        let key = storage::bytes_to_key(KeyPrefix::CrossContractCall, WNEAR_KEY);
+        match io.read_storage(&key) {
+            Some(bytes) => Address::try_from_slice(&bytes.to_vec()).expect(ERR_CORRUPTED_STORAGE),
+            None => panic!("{}", ERR_MISSING_WNEAR_ADDRESS),
+        }
+    }
+
+    /// Get the latest router contract version.
+    pub fn get_latest_code_version<I: IO>(io: &I) -> CodeVersion {
+        let key = storage::bytes_to_key(KeyPrefix::CrossContractCall, VERSION_KEY);
+        read_version(io, &key).unwrap_or_default()
+    }
+
+    /// Get the version of the currently deploy router for the given address (if it exists).
+    pub fn get_code_version_of_address<I: IO>(io: &I, address: &Address) -> Option<CodeVersion> {
+        let key = storage::bytes_to_key(KeyPrefix::CrossContractCall, address.as_bytes());
+        read_version(io, &key)
+    }
+
+    /// Private utility method for reading code version from storage.
+    fn read_version<I: IO>(io: &I, key: &[u8]) -> Option<CodeVersion> {
+        match io.read_u32(key) {
+            Ok(value) => Some(CodeVersion(value)),
+            Err(ReadU32Error::MissingValue) => None,
+            Err(ReadU32Error::InvalidU32) => panic!("{}", ERR_CORRUPTED_STORAGE),
+        }
+    }
+}
+
+fn transfer_from_args(from: H160, to: H160, amount: U256) -> Vec<u8> {
+    let args = ethabi::encode(&[
+        ethabi::Token::Address(from),
+        ethabi::Token::Address(to),
+        ethabi::Token::Uint(amount),
+    ]);
+    [&consts::TRANSFER_FROM_SELECTOR, args.as_slice()].concat()
+}
+
 fn create_target_account_id(sender: H160, engine_account_id: &str) -> AccountId {
     format!("{}.{}", hex::encode(sender.as_bytes()), engine_account_id)
         .parse()
         .unwrap()
 }
 
+fn revert_with_message(message: &str) -> PrecompileFailure {
+    PrecompileFailure::Revert {
+        exit_status: evm::ExitRevert::Reverted,
+        output: message.as_bytes().to_vec(),
+    }
+}
```

Comment:
We should use this for other precompiles. Possibly as a macro, with a default message if no message is set.

Reply:
I think this should be out of scope for this PR because it is large enough as is. Please file an issue and the engine team can take on that refactor as a follow up PR after the xcc feature is merged.

Reply:
Do make note in the description that a follow up PR should follow something like that, then on some clean up cycle we can make it the same.